File: packages/expo-image-manipulator/android/src/main/java/expo/modules/imagemanipulator/arguments/ActionCrop.java
package expo.modules.imagemanipulator.arguments;

import androidx.annotation.NonNull;

import java.util.Map;

public class ActionCrop {
  private static final String TAG = "action.crop";

  private static final String KEY_ORIGIN_X = "originX";
  private static final String KEY_ORIGIN_Y = "originY";
  private static final String KEY_WIDTH = "width";
  private static final String KEY_HEIGHT = "height";

  @NonNull
  private final Integer mOriginX;
  @NonNull
  private final Integer mOriginY;
  @NonNull
  private final Integer mWidth;
  @NonNull
  private final Integer mHeight;

  private ActionCrop(
      @NonNull Integer originX,
      @NonNull Integer originY,
      @NonNull Integer width,
      @NonNull Integer height) {
    mOriginX = originX;
    mOriginY = originY;
    mWidth = width;
    mHeight = height;
  }

  static ActionCrop fromObject(Object options) throws IllegalArgumentException {
    if (!(options instanceof Map<?,?>)){
      throw new IllegalArgumentException("'" + TAG + "' must be an object");
    }
    Map optionsMap = (Map) options;

    Double originXDouble = ActionCrop.getDoubleFromOptions(optionsMap, KEY_ORIGIN_X, TAG + "." + KEY_ORIGIN_X);
    if (originXDouble == null) {
      throw new IllegalArgumentException("'" + TAG + "." + KEY_ORIGIN_X + "' must be defined");
    }
    Integer originX = originXDouble.intValue();

    Double originYDouble = ActionCrop.getDoubleFromOptions(optionsMap, KEY_ORIGIN_Y, TAG + "." + KEY_ORIGIN_Y);
    if (originYDouble == null) {
      throw new IllegalArgumentException("'" + TAG + "." + KEY_ORIGIN_Y + "' must be defined");
    }
    Integer originY = originYDouble.intValue();

    Double widthDouble = ActionCrop.getDoubleFromOptions(optionsMap, KEY_WIDTH, TAG + "." + KEY_WIDTH);
    if (widthDouble == null) {
      throw new IllegalArgumentException("'" + TAG + "." + KEY_WIDTH + "' must be defined");
    }
    Integer width = widthDouble.intValue();

    Double heightDouble = ActionCrop.getDoubleFromOptions(optionsMap, KEY_HEIGHT, TAG + "." + KEY_HEIGHT);
    if (heightDouble == null) {
      throw new IllegalArgumentException("'" + TAG + "." + KEY_HEIGHT + "' must be defined");
    }
    Integer height = heightDouble.intValue();

    return new ActionCrop(originX, originY, width, height);
  }

  private static Double getDoubleFromOptions(Map options, String key, String pathToErroneousKey) {
    if (!options.containsKey(key)) {
      return null;
    }
    if (!(options.get(key) instanceof Double)) {
      throw new IllegalArgumentException("'" + pathToErroneousKey + "' must be a Number value");
    }
    return ((Double) options.get(key));
  }

  @NonNull
  public Integer getOriginX() {
    return mOriginX;
  }

  @NonNull
  public Integer getOriginY() {
    return mOriginY;
  }

  @NonNull
  public Integer getWidth() {
    return mWidth;
  }

  @NonNull
  public Integer getHeight() {
    return mHeight;
  }
}

File: packages/expo-image-manipulator/android/src/main/java/expo/modules/imagemanipulator/arguments/ActionCrop.kt
package expo.modules.imagemanipulator.arguments

class ActionCrop private constructor(
  val originX: Int,
  val originY: Int,
  val width: Int,
  val height: Int) {

  companion object {
    private const val TAG = "action.crop"
    private const val KEY_ORIGIN_X = "originX"
    private const val KEY_ORIGIN_Y = "originY"
    private const val KEY_WIDTH = "width"
    private const val KEY_HEIGHT = "height"
    @Throws(IllegalArgumentException::class)
    fun fromObject(options: Any?): ActionCrop {
      require(options is Map<*, *>) { "'$TAG' must be an object" }
      val optionsMap = options
      val originXDouble = getDoubleFromOptions(optionsMap, KEY_ORIGIN_X, TAG + "." + KEY_ORIGIN_X)
        ?: throw IllegalArgumentException("'" + TAG + "." + KEY_ORIGIN_X + "' must be defined")
      val originX = originXDouble.toInt()
      val originYDouble = getDoubleFromOptions(optionsMap, KEY_ORIGIN_Y, TAG + "." + KEY_ORIGIN_Y)
        ?: throw IllegalArgumentException("'" + TAG + "." + KEY_ORIGIN_Y + "' must be defined")
      val originY = originYDouble.toInt()
      val widthDouble = getDoubleFromOptions(optionsMap, KEY_WIDTH, TAG + "." + KEY_WIDTH)
        ?: throw IllegalArgumentException("'" + TAG + "." + KEY_WIDTH + "' must be defined")
      val width = widthDouble.toInt()
      val heightDouble = getDoubleFromOptions(optionsMap, KEY_HEIGHT, TAG + "." + KEY_HEIGHT)
        ?: throw IllegalArgumentException("'" + TAG + "." + KEY_HEIGHT + "' must be defined")
      val height = heightDouble.toInt()
      return ActionCrop(originX, originY, width, height)
    }

    private fun getDoubleFromOptions(options: Map<*, *>, key: String, pathToErroneousKey: String): Double? {
      if (!options.containsKey(key)) {
        return null
      }
      require(options[key] is Double) { "'$pathToErroneousKey' must be a Number value" }
      return options[key] as Double?
    }
  }
}
